### PR TITLE
feat(cli): add --staged flag

### LINKED ISSUE #2296: 📎 Feature: --staged option to simplify the creation of git hook scripts
## Problem Statement

As of today, `biome` does not provide a direct mechanism to select (for linting, formatting or checking) only the files that are staged for a commit.

We have a "similar" flag, `--changed`, but its main purpose is to be used in CI, and allows us to check only the files that changed between 2 points in history (which can be explicit or implicit).

Many people think at first that `--changed` will also help them to select only the files that are staged, not being the case.

More advanced users have found workarounds to this issue ( https://github.com/biomejs/biome/issues/1198#issuecomment-1858364218 ), but these involve having to write longer, more complex and slower scripts. Not everyone can come up with these solutions, even if they seem simple after we see them.

Others rely on tools such as `lint-staged`, but this involves adding more dependencies to the project (which in this case are much slower than `biome`), and opens up the possibility for more bugs caused by unexpected interactions between `lint-staged` and `biome`.

## Proposal

After a discussion held in Discord ( https://discord.com/channels/1132231889290285117/1222450681919832134 ), we agreed that it would be a good idea to introduce a new `--staged` flag for the `lint`, `format` and `check` commands that will make biome take into account only the files that are _staged_.

### Addendum

While `--staged` doesn't give us much when we are running "potentially destructive" commands such as `format` or `check --apply` (because we still need to independently obtain a list of staged files so we can re-add them to the "stage" before they are committed), it gives a clear advantage for readonly-mode commands such as `lint` without `--apply`.

### PR DESCRIPTION
**Pull request by castarco**
## Summary

This PR introduces a new `--staged` flag that allows `biome` to select only the files that are staged to be committed.

#### Long version

- Fixes #2296
- Related Discord discussion: https://discord.com/channels/1132231889290285117/1222450681919832134

#### TLDR

There is some semblance with `--changed`, but its behaviour is different:
- `--changed` is useful to run certain checks in CI environment, selecting only the files that changed between two git "refs" (they can be explicit or implicit).
- In contrast, `--staged` is useful in local environments, to be used in git hooks such as `pre-commit`.

## Test Plan

- I introduced new unit tests covering basic behaviour changes
- I tested the generated binary manually to verify that it works exactly as expected

## "Mixed" Performance Results

What follows is for a very small repository.

### When there are no changes

Performance is slightly better

Screenshot 2024-04-04 at 14 28 49

### When there is a small number of changes 

Performance is slightly worse 🤔, I suspect that this because of the extra `git` call.

Screenshot 2024-04-04 at 14 28 28

I hypothesize that this extra overhead should be "absorbed" in bigger repositories with many more files.

### CODE REVIEW

**Review by ematipico — COMMENTED** [review]
I had a quick look, and it seems that the CI command isn't covered? 

I know that the new option shouldn't work in CI, but we should at least terminate the process and raise a diagnostic, saying that this option isn't valid in the CI environments

File: crates/biome_cli/src/commands/format.rs [hunk 1]
```diff
@@ -157,12 +159,22 @@ pub(crate) fn format(
...
     }
 
     if changed {
+        if staged {
+            return Err(CliDiagnostic::incompatible_arguments("changed", "staged"));
+        }
         paths = get_changed_files(&session.app.fs, &configuration, since)?;
+    } else if staged {
+        paths = get_staged_files(&session.app.fs)?;
     }
```
Comment (ematipico) [inline crates/biome_cli/src/commands/format.rs:178]: Considering how the code is repeated, maybe it makes more sense to have a function in `commands/mod.rs` that does the check, and each command imports it and calls it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -91,6 +91,12 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
 
   Contributed by @Conaclos
 
+- Added new `--staged` flag to the `check`, `format` and `lint` subcommands.
+
+  This new option allows users to apply the command _only_ to the files that are staged (the
+  ones that will be committed), which can be very useful to simplify writting git hook scripts
+  such as `pre-commit`.
```
Comment (ematipico) [inline CHANGELOG.md:98]:
```suggestion
- Added new `--staged` flag to the `check`, `format` and `lint` subcommands.

  This new option allows users to apply the command _only_ to the files that are staged (the
  ones that will be committed), which can be very useful to simplify writing git hook scripts
  such as `pre-commit`.
```

File: website/src/content/docs/internals/changelog.md [hunk 3]
```diff
@@ -97,6 +97,12 @@ our [guidelines for writing a good changelog entry](https://github.com/biomejs/b
 
   Contributed by @Conaclos
 
+- Added new `--staged` flag to the `check`, `format` and `lint` subcommands.
+
+  This new option allows users to apply the command _only_ to the files that are staged (the
+  ones that will be committed), which can be very useful to simplify writting git hook scripts
```
Comment (ematipico) [inline website/src/content/docs/internals/changelog.md:103]:
```suggestion
  ones that will be committed), which can be very useful to simplify writing git hook scripts
```

**Review by ematipico — APPROVED** [review]
I only noticed now that the new options were added only for specific commands, so you can discard my previous comment. I left few suggestions, and I think we can merge the PR once they are addressed

File: crates/biome_cli/src/commands/format.rs (hunk 1 above)
Comment (castarco) [reply crates/biome_cli/src/commands/format.rs]: Extracted 🏗️ .

**Review by ematipico — APPROVED** [review]
This is a great addition; thank you!

**Comment by Conaclos** [comment]
While I was writing the blog post for the Biome 1.7, I wonder if we should not rename `--staged` to `--staged-files` to highlight the fact that unstated changes in staged files are not ignored.

Any opinion?

**Comment by ematipico** [comment]
> highlight the fact that unstated changes in staged files are not ignored.

I don't understand what you mean. I think `--staged` is fine; it's the git dialect.

**Comment by Conaclos** [comment]
> I don't understand what you mean. I think --staged is fine; it's the git dialect.

i mean that you can stage a file and then modify it:

```shell
❯ git status
On branch main
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   a

❯ echo "more" >> a

❯ git status
On branch main
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   a

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   a
```

You end with a staged file with unstaged changes.

**Comment by castarco** [comment]
> You end with a staged file with unstaged changes.

@Conaclos That's true, but there are some relevant points:
- many people run their hooks as linters without `--apply`.
- Without staged, when we perform the git command to obtain the staged files, many times they filter them by extension. This can be a problem as it is, in a way, having multiple sources of truth (we have "include/exclude" filters in Biome's configuration, but also in our shell script)
- This functionality can be further extended in the future, in two separate ways (not necessarily exclusive, but if any of them is implemented, then the other becomes irrelevant):
  - Making it possible for the `biome format` and `biome check` commands to output only the list of files they changed. This could be passed directly as arguments to `git add`.
  - Internalizing the `git add` command inside Biome (although I prefer the first one, as it is lighter, and more generic)

**Comment by Conaclos** [comment]
@castarco 
I didn't suggest fixing it, just renaming the flag to be clear about that.
Any way I am also happy with `--staged`. I was just expressing a concern/idea.

> For this I imagine 2 potential solutions:
>
>    When there is a discrepancy between "staged" and "local filesystem", Biome could issue a warning.
>        This one requires an extra git command to obtain the list of unstaged changes.
>    A more sophisticated and aggressive approach would involve running a pair of git stash and git stash pop commands (before and after the fixes).
        This one requires 2 extra git commands (and they are also slower because they imply filesystem writes)

The first option might be nice :)
This seems to be a good fit for Biome.

I think the second suggestion is a bad idea, because if the user uses `--apply`/`--write`, you can end up with conflicts when reapplying the stashed changes.
This can be very confusing for users.
This sort of thing should be left to dedicated tools like [git-format-staged](https://biomejs.dev/recipes/git-hooks/#git-format-staged) and the like.

> Without staged, when we perform the git command to obtain the staged files, many times "we" filter them by extension before passing them to Biome. This can be a problem because, in a way, it is equivalent to having multiple sources of truth (we have "include/exclude" filters in Biome's configuration, but also in our shell script)

I am unsure to follow you there. Users should rely only on a Biome configuration and `--files-ignore-unknown=true` to filter files.

> Making it possible for the biome format and biome check commands to output only the list of files they changed. This could be passed directly as arguments to git add.

I am not sure if there is so much value for that. A user could simply run `git update-index --again` — assuming that there is no unstaged changes before the formatting/linting.
However, I am not opposed to the idea, it could have other applications...

> Internalizing the git add command inside Biome (although I prefer the previous option over this one, as it is lighter and more generic)

I think that it is a bad idea because this could be confusing and blur the line between Biome and Git.
